### ISSUE TITLE
i2c type alias

### ISSUE THREAD

**Issue by pdgilbert** [stm32-rs/stm32f3xx-hal#288]
(Related to #214) I am trying to define a type alias for an i2c interface because I need it in `rtic` `struct Shared`. I've tried several variations with no success. For example, `type I2cBus = I2c<I2C1, (PB6<Alternate<OpenDrain, AF4>>, PB7<Alternate<OpenDrain, AF4>>)>` gives
```
  --> examples/zzot.rs:44:9
   |
44 |         pac,
   |         ^^^
...
47 |         serial::{RxPin, Serial, Tx, TxPin},
   |                  ^^^^^
   |
   = note: `#[warn(unused_imports)]` on by default

error[E0747]: type provided when a constant was expected
  --> examples/zzot.rs:57:55
   |
57 |     type I2cBus = I2c<I2C1, (PB6<Alternate<OpenDrain, AF4>>, PB7<Alternate<OpenDrain, AF4>>)>;
   |                                                       ^^^
   |
help: if this generic argument was intended as a const parameter, surround it with braces
   |
57 |     type I2cBus = I2c<I2C1, (PB6<Alternate<OpenDrain, { AF4 }>>, PB7<Alternate<OpenDrain, AF4>>)>;
   |                                                       ^     ^

error[E0747]: type provided when a constant was expected
  --> examples/zzot.rs:57:87
   |
57 |     type I2cBus = I2c<I2C1, (PB6<Alternate<OpenDrain, AF4>>, PB7<Alternate<OpenDrain, AF4>>)>;
   |                                                                                       ^^^
   |
help: if this generic argument was intended as a const parameter, surround it with braces
   |
57 |     type I2cBus = I2c<I2C1, (PB6<Alternate<OpenDrain, AF4>>, PB7<Alternate<OpenDrain, { AF4 }>>)>;
   |                                                                                       ^     ^

error: aborting due to 2 previous errors; 1 warning emitted
```
but surrounding in brackets makes things worse.  (I'm using git `stm32f3xx-hal#53567500d7ab0`. With different `setup()` functions this example works with several other device hals.)  Suggestions would be appreciated. 

The complete example code is 
<details>
<summary>Click to expand</summary>

```
#![deny(unsafe_code)]
#![no_std]
#![no_main]

#[cfg(debug_assertions)]
use panic_semihosting as _;

#[cfg(not(debug_assertions))]
use panic_halt as _;
//use panic_rtt_target as _;

use rtic::app;

#[cfg_attr(feature = "stm32f3xx", app(device = stm32f3xx_hal::pac,   dispatchers = [TIM2, TIM3]))] 

mod app {

    use dwt_systick_monotonic::DwtSystick;
    use rtic::time::duration::{Seconds, };

    use cortex_m::asm; //asm::delay(N:u32) blocks the program for at least N CPU cycles.

    use core::fmt::Write;
    use embedded_ccs811::{
        mode as Ccs811Mode, prelude::*, AlgorithmResult, Ccs811Awake, MeasurementMode,
        SlaveAddr as Ccs811SlaveAddr,
    };
    use hdc20xx::{mode as Hdc20xxMode, Hdc20xx, SlaveAddr as Hdc20xxSlaveAddr};
    use nb::block;
    use rtt_target::{rprintln, rtt_init_print};
    use shared_bus_rtic::SharedBus;
    
    const PERIOD: Seconds = Seconds(10);
   

    #[cfg(feature = "stm32f3xx")] //  eg Discovery-stm32f303
    use stm32f3xx_hal::{
        gpio::{Alternate, OpenDrain, Output, PushPull, AF4,
            gpiob::{PB6, PB7},
            gpioe::PE9,
         //   Otyper::Otype,      
        },
        i2c::{I2c, SclPin, SdaPin},
        pac,
        pac::{Peripherals, I2C1, USART1},
        prelude::*,
        serial::{RxPin, Serial, Tx, TxPin},
    };
       
    #[cfg(feature = "stm32f3xx")]
    const CLOCK: u32 = 8_000_000; //should be set for board not for HAL

    #[cfg(feature = "stm32f3xx")]
    type LedType = PE9<Output<PushPull>>;
    
    #[cfg(feature = "stm32f3xx")]
    type I2cBus = I2c<I2C1, (PB6<Alternate<OpenDrain, AF4>>, PB7<Alternate<OpenDrain, AF4>>)>;

//    type I2cBus = I2c<I2C1, (impl SclPin<I2C1>, impl SdaPin<I2C1>)>;  //`impl Trait` in type aliases is unstable
//    type I2cBus = I2c<I2C1, (PB6<Alternate<AF4<OpenDrain, AF4>>>, PB7<Alternate<AF4<OpenDrain, AF4>>>)>;
//    type I2cBus = I2c<I2C1, (PB6<Alternate<OpenDrain, AF4>>, PB7<Alternate<OpenDrain, AF4>>)>;
//    type I2cBus = I2c<I2C1, (PB6<Alternate<AF4<Otype>, OpenDrain>>, PB7<Alternate<AF4<Otype>, OpenDrain>>)>;
//    type I2cBus = I2c<I2C1, (PB6<Alternate<AF4, OpenDrain>>, PB7<Alternate<AF4, OpenDrain>>)>;
//    type I2cBus = I2c<I2C1, (PB6<Alternate<<OpenDrain>, 4_u8>>, PB7<Alternate<<OpenDrain>, 4_u8>>)>;

    //type I2cBus = I2c<pac::I2C1, (PB6<Alternate<Otype, AF4>>, PB7<Alternate<Otype, AF4>>)>;
       
    #[cfg(feature = "stm32f3xx")]
    type TxType = Tx<USART1, dyn TxPin<USART1>>;
    //type TxType = Tx<USART1, impl TxPin<USART1>>;

    #[cfg(feature = "stm32f3xx")]
    fn setup(dp: Peripherals) -> (I2c<I2C1, (impl SclPin<I2C1>, impl SdaPin<I2C1>)>, LedType, TxType ) {
       //fn setup(dp: Peripherals) -> (I2cBus, LedType, TxType ) {
        let mut flash = dp.FLASH.constrain();
        let mut rcc = dp.RCC.constrain();
        let clocks = rcc.cfgr.freeze(&mut flash.acr);
    
        let mut gpiob = dp.GPIOB.split(&mut rcc.ahb);
        let scl = gpiob
            .pb6
            .into_af4_open_drain(&mut gpiob.moder, &mut gpiob.otyper, &mut gpiob.afrl);
        let sda = gpiob
            .pb7
            .into_af4_open_drain(&mut gpiob.moder, &mut gpiob.otyper, &mut gpiob.afrl);
        let i2c = I2c::new(dp.I2C1, (scl, sda), 100_000.Hz(), clocks, &mut rcc.apb1);
  
        let mut gpioa = dp.GPIOA.split(&mut rcc.ahb);
        let (tx, _rx) = Serial::usart1(
            dp.USART1,
            (
                gpioa
                    .pa9
                    .into_af7_push_pull(&mut gpioa.moder, &mut gpioa.otyper, &mut gpioa.afrh), //tx pa9
                gpioa
                    .pa10
                    .into_af7_push_pull(&mut gpioa.moder, &mut gpioa.otyper, &mut gpioa.afrh), //rx pa10
            ),
            115200.Bd(),
            clocks,
            &mut rcc.apb2,
        )
        .split();
    
        let mut gpioe = dp.GPIOE.split(&mut rcc.ahb);
        let led = gpioe
            .pe9
            .into_push_pull_output(&mut gpioe.moder, &mut gpioe.otyper);

        impl LED for PE9<Output<PushPull>> {
            fn on(&mut self) -> () {
                self.set_high().unwrap()
            }
            fn off(&mut self) -> () {
                self.set_low().unwrap()
            }
        }
        led.off();
    
        (i2c, led, tx)
    }
        
    pub trait LED {
        // depending on board wiring, on may be set_high or set_low, with off also reversed
        // implementation should deal with this difference
        fn on(&mut self) -> ();
        fn off(&mut self) -> ();
    }
    
    #[monotonic(binds = SysTick, default = true)]
    type DwtMono = DwtSystick<CLOCK>;
    
    /*
     * shared-bus-rtic aggregate: multiple peripherals on a single i2c bus
     *
     * According to shared-bus-rtic docs:
     * Note that all of the drivers that use the same underlying bus **must** be stored within a single
     * resource (e.g. as one larger `struct`) within the RTIC resources. This ensures that RTIC will
     * prevent one driver from interrupting another while they are using the same underlying bus.
     */
        
       
    #[shared]
    struct Shared {
        led: LedType,
        ccs811: Ccs811Awake<SharedBus<I2cBus>, Ccs811Mode::App>,
        hdc2080: Hdc20xx<SharedBus<I2cBus>, Hdc20xxMode::OneShot>,
        tx : TxType,
    }
    
    #[local]
    struct Local {
       led_state: bool, 
       env: [(f32, f32); 1200],
       index: usize,
       measurements: [AlgorithmResult; 1200],
    }

    #[init]
    fn init(mut cx: init::Context) -> (Shared, Local, init::Monotonics) {
        let mono = DwtSystick::new(&mut cx.core.DCB, cx.core.DWT, cx.core.SYST, CLOCK);

        rtt_init_print!();
        rprintln!("CCS811/HDC2080 example");

        let device: Peripherals = cx.device;

        let (i2c, mut led, mut tx) = setup(device);

        led.off();

        // Previously these were initialized static mut in fn measure()
        let led_state: bool;
        let env: [(f32, f32); 1200] = [(0.0, 0.0); 1200];
        let index: usize = 0;
        let measurements:[AlgorithmResult; 1200] =
             [AlgorithmResult { eco2: 0, etvoc: 0, raw_current: 0, raw_voltage: 0, }; 1200];

        let manager = shared_bus_rtic::new!(i2c, I2cBus);
        let mut hdc2080 = Hdc20xx::new(manager.acquire(), Hdc20xxSlaveAddr::default());
        let mut ccs811 = Ccs811Awake::new(manager.acquire(), Ccs811SlaveAddr::default());
        ccs811.software_reset().unwrap();
        
        // Delay while ccs811 resets.
        // Note that this delay cannot use SYST because Monotonics uses that (for spawn,
        //   although spawn has not yet happened so there may be a way?)
        
        led.on();
        asm::delay(3 * CLOCK); // (3 * CLOCK cycles give aprox 3 second delay
        //delay::delay_ms(3_000_u16);
        led.off();
        led_state = false;

        let mut ccs811 = ccs811.start_application().ok().unwrap();
        let en = block!(hdc2080.read()).unwrap();
        ccs811
            .set_environment(en.temperature, en.humidity.unwrap_or(0.0))
            .unwrap();
        ccs811.set_mode(MeasurementMode::ConstantPower1s).unwrap();

        //cx.schedule.measure(cx.start + PERIOD.cycles()).unwrap();
        measure::spawn_after(PERIOD).unwrap();

        writeln!(tx, "start\r",).unwrap();

        (Shared {led: led, ccs811: ccs811, hdc2080: hdc2080, tx: tx }, 
         Local {led_state: led_state,  env: env,  index: index,  measurements: measurements}, 
         init::Monotonics(mono))
     }

    #[task(shared = [led, ccs811, hdc2080, tx], local = [led_state, env, index, measurements])]
    fn measure(mut cx: measure::Context) {

        if *cx.local.led_state {
            cx.shared.led.lock(|led| led.off());
            *cx.local.led_state = false;
        } else {
            cx.shared.led.lock(|led| led.on());
            *cx.local.led_state = true;
        }

        let default = AlgorithmResult::default();
        if *cx.local.index < cx.local.measurements.len() {
            //let data = block!(cx.shared.ccs811.data()).unwrap_or(default);
            let data = cx.shared.ccs811.lock(|ccs811| block!(ccs811.data())).unwrap_or(default);
            cx.local.measurements[*cx.local.index] = data;
            //let en = block!(cx.shared.hdc2080.read()).unwrap();
            let en = cx.shared.hdc2080.lock(|hdc2080| block!(hdc2080.read())).unwrap();
            let temp = en.temperature;
            let humidity = en.humidity.unwrap_or(0.0);
            cx.local.env[*cx.local.index] = (temp, humidity);
            *cx.local.index += 1;
            //cx.shared.ccs811.set_environment(temp, humidity).unwrap();
            cx.shared.ccs811.lock(|ccs811| ccs811.set_environment(temp, humidity)).unwrap();
        }
        //writeln!(cx.shared.tx, "\rstart\r",).unwrap();
        cx.shared.tx.lock(|tx| writeln!( tx, "\rstart\r",)).unwrap();
        for i in 0..*cx.local.index {
            let data = cx.local.measurements[i];
            let en = if i == 0 { (0.0, 0.0) } else { cx.local.env[i - 1] };
            //writeln!(cx.shared.tx,  "{},{},{},{},{},{:.2},{:.2}\r", i, data.eco2, 
            //        data.etvoc, data.raw_current, data.raw_voltage, en.0, en.1 ).unwrap();
            cx.shared.tx.lock(|tx| writeln!(tx,  "{},{},{},{},{},{:.2},{:.2}\r", i, data.eco2, 
                    data.etvoc, data.raw_current, data.raw_voltage, en.0, en.1 )).unwrap();
        }
        //cx.schedule.measure(cx.scheduled + PERIOD.cycles()).unwrap();
        measure::spawn_after(PERIOD).unwrap();
    }

}
```
</details>

**Comment by Sh3Rm4n** [comment]
Without checking, either 

```rust


type I2cBus = I2c<I2C1, (PB6<Alternate<OpenDrain, { 4_u8 }>>, PB7<Alternate<OpenDrain, { 4_u8 }>>)>;
```

or 

```rust
type I2cBus = I2c<I2C1, (PB6<AF4<OpenDrain>>, PB7<AF4<OpenDrain>>)>;
```

should work.

[`AF4`](https://docs.rs/stm32f3xx-hal/0.8.0/stm32f3xx_hal/gpio/type.AF4.html) in itself is a type alias, which hides the fact, that const generics are used to define  the type.

I know the compiler gives some misleading (or not easy to understand) suggestions on these highly generic types, so only relying on the compiler sadly leads to the trial and error you stepped into (I have myself in the past, so this part needs to definitely be improved, most probably on stm32f3xx-hal's side) 

As you can see [`SclPin`](https://docs.rs/stm32f3xx-hal/0.8.0/stm32f3xx_hal/i2c/trait.SclPin.html) for example has some [implementors](https://docs.rs/stm32f3xx-hal/0.8.0/stm32f3xx_hal/i2c/trait.SclPin.html#implementors) 

![grafik](https://user-images.githubusercontent.com/29653149/135834027-792bc32c-4185-4f32-b3ef-8a5b3d339031.png)

So choosing one of these for the type alias should give you the quickest success.

**Comment by pdgilbert** [comment]
Thanks @Sh3Rm4n, both of those work. Also thanks for the pointers on finding type definitions. I've been down the rabbit hole of following compiler suggestions several times. I'm now having a slightly different problem with `doesn't have a size known at compile-time` for the TxType.  Adding `dyn` or not just removes a warning.  Any suggestins would be appreciated. I'm getting

```
Compiling rust-integration-testing-of-examples v0.1.0 (/home/paul/githubClones/rust-integration-testing/dev-testing)
error[E0277]: the size for values of type `(dyn stm32f3xx_hal::serial::TxPin<USART1> + 'static)` cannot be known at compilation time
   --> examples/zzot.rs:63:34
    |
63  |     fn setup(dp: Peripherals) -> (I2cBus, LedType, TxType ) {
    |                                  ^^^^^^^^^^^^^^^^^^^^^^^^^^ doesn't have a size known at compile-time
    | 
   ::: /home/paul/.cargo/git/checkouts/stm32f3xx-hal-572339a820137a38/5356750/src/serial.rs:346:26
    |
346 |     pub struct Tx<Usart, Pin> {
    |                          --- required by this bound in `stm32f3xx_hal::serial::Tx`
    |
    = help: the trait `Sized` is not implemented for `(dyn stm32f3xx_hal::serial::TxPin<USART1> + 'static)`

error[E0277]: the size for values of type `(dyn stm32f3xx_hal::serial::TxPin<USART1> + 'static)` cannot be known at compilation time
   --> examples/zzot.rs:137:14
    |
137 |         tx : TxType,
    |              ^^^^^^ doesn't have a size known at compile-time
    | 
   ::: /home/paul/.cargo/git/checkouts/stm32f3xx-hal-572339a820137a38/5356750/src/serial.rs:346:26
    |
346 |     pub struct Tx<Usart, Pin> {
    |                          --- required by this bound in `stm32f3xx_hal::serial::Tx`
    |
    = help: the trait `Sized` is not implemented for `(dyn stm32f3xx_hal::serial::TxPin<USART1> + 'static)`

error[E0277]: the size for values of type `(dyn stm32f3xx_hal::serial::TxPin<USART1> + 'static)` cannot be known at compilation time
   --> examples/zzot.rs:14:35
    |
14  | #[cfg_attr(feature = "stm32f3xx", app(device = stm32f3xx_hal::pac,   dispatchers = [TIM2, TIM3]))] 
    |                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ doesn't have a size known at compile-time
    | 
   ::: /home/paul/.cargo/git/checkouts/stm32f3xx-hal-572339a820137a38/5356750/src/serial.rs:346:26
    |
346 |     pub struct Tx<Usart, Pin> {
    |                          --- required by this bound in `stm32f3xx_hal::serial::Tx`
    |
    = help: the trait `Sized` is not implemented for `(dyn stm32f3xx_hal::serial::TxPin<USART1> + 'static)`
    = note: this error originates in an attribute macro (in Nightly builds, run with -Z macro-backtrace for more info)

error: aborting due to 3 previous errors


```

**Comment by Sh3Rm4n** [comment]
You are using 

```rust
type TxType = Tx<USART1, dyn TxPin<USART1>>;
```

which internally already uses `dyn`, which is `!Sized` (unsized) by default (IIRC) so you either have to `Box` it, which is not possible in `no_std` or use `&dyn TxPin`. **However**, I think you do not need to reach for dynamic dispatch here, what you are trying to do, is to create a type alias for a **concrete** type you already know at this point. 

Looking at the definition of [`Tx`](https://docs.rs/stm32f3xx-hal/0.8.0/stm32f3xx_hal/serial/struct.Tx.html#impl) you have to provide a [`serial::Instance`](https://docs.rs/stm32f3xx-hal/0.8.0/stm32f3xx_hal/serial/trait.Instance.html), which you have done already ([`USART1`](https://docs.rs/stm32f3xx-hal/0.8.0/stm32f3xx_hal/pac/struct.USART1.html)) **and** have to provide a pin, which implements [`TxPin`](https://docs.rs/stm32f3xx-hal/0.8.0/stm32f3xx_hal/serial/trait.TxPin.html). 

Looking at your code, you are trying to use PA9: 

```rust
        let (tx, _rx) = Serial::usart1(
            dp.USART1,
            (
                gpioa
                    .pa9
                    .into_af7_push_pull(&mut gpioa.moder, &mut gpioa.otyper, &mut gpioa.afrh), //tx pa9
                gpioa
                    .pa10
                    .into_af7_push_pull(&mut gpioa.moder, &mut gpioa.otyper, &mut gpioa.afrh), //rx pa10
            ),
            115200.Bd(),
            clocks,
            &mut rcc.apb2,
        )
        .split();
```

which is implementing `TxPin`: [`impl<Otype> TxPin<USART1> for PA9<AF7<Otype>>`](https://docs.rs/stm32f3xx-hal/0.8.0/stm32f3xx_hal/serial/trait.TxPin.html#impl-TxPin%3CUSART1%3E). 

So the complete type is: 

```rust
type TxType = Tx<USART1, PA9<AF7<OpenDrain>>>;
```

**Comment by pdgilbert** [comment]
Thank you again @Sh3Rm4n. Indeed I was not trying to do dynamic dispatch and only got there by following compiler suggestions. The `type TxType = Tx<USART1, PA9<AF7<OpenDrain>>>;` works but I am curious why I do not have to specify the pin with other hals. These all work:
```
    #[cfg(feature = "stm32f1xx")]
    type TxType = Tx<USART1>;

    #[cfg(feature = "stm32f4xx")]
    type TxType = Tx<USART1>;

    #[cfg(feature = "stm32f7xx")]
    type TxType = Tx<pac::USART2>;

   #[cfg(feature = "stm32h7xx")]
    type TxType = Tx<USART2>;

    #[cfg(feature = "stm32l1xx")]
    type TxType = Tx<USART1>;

    #[cfg(feature = "stm32l4xx")]
    type TxType = Tx<USART2>;
```

**Comment by Sh3Rm4n** [comment]
This is because in other HALs, the [`split()`](https://docs.rs/stm32f4xx-hal/0.10.0/stm32f4xx_hal/serial/struct.Serial.html#method.split) does return a `Tx` type, which looses the information, from what pin it was created. 

For the `stm32f3xx-hal`, this information is kept: See [`Serial::split()`](https://docs.rs/stm32f3xx-hal/0.8.0/stm32f3xx_hal/serial/struct.Serial.html#method.split).

This was necessary it to be able to implement [`Serial::join()`](https://docs.rs/stm32f3xx-hal/0.8.0/stm32f3xx_hal/serial/struct.Serial.html#method.join) implemented in #252. This functionality is not provided by the other HALs.

**Comment by pdgilbert** [comment]
That makes sense. At this rate I might actual start to understand these interfaces someday. Thanks for the explanation.
